django-data/image/common/helpers.py

```diff
@@ -72,3 +74,15 @@ def format_callback(obj):
             len(objs) for model, objs in collector.model_objs.items()}
 
     return to_delete, model_count, protected
+
+
+async def send_message_to_websocket(message, pk):
+    """
+    Function will create websocket object and send message to django-channels
+    :param message: message to send to websocket
+    :param pk: primary key of submission
+    """
+    # Need to have it here as in case with small test data message sent to websocket will overcome response from server
+    time.sleep(3)
```

Comment:
Wrote tests for submissions_ws itself, next thing is to fix existing unittests to use mock instead of real function

Reply:
Ty. Tomorrow I could give you AN example

Reply:
Hi Alexey, 

there are two ways to mocking object in your case. I'm not an expert in mocking channels, but if you take [test_task.py](https://github.com/cnr-ibba/IMAGE-InjectTool/blob/b10ceb20b08eb5009cd29af59265900c3a162508/django-data/image/biosample/tests/test_tasks.py#L107), you can see an example of I mock object using patch decorator. This is patched only in this test. Outside of this, the function will called normally as the site does. Another way to mocking object, could be done like this
```python
class SubmitTestCase(SubmitMixin, RedisMixin, TestCase):

    @classmethod
    def setUpClass(cls):
        # calling my base class setup
        super().setUpClass()

        # starting mocked objects
        cls.mock_ws_patcher = patch(
            'asyncio.get_event_loop')
        cls.mock_ws = cls.mock_ws_patcher.start()

    @classmethod
    def tearDownClass(cls):
        # stopping mocked object
        cls.mock_ws_patcher.stop()

        # calling base method
        super().tearDownClass()

    def setUp(self):
        # call Mixin method
        super().setUp()

        # setting tasks
        self.my_task = SubmitTask()
```
With this code, you can execute `docker-compose run --rm uwsgi pytest --verbosity=2 biosample/tests/test_tasks.py::SubmitTestCase` without using redis and channel (check asgi log with docker logs)
By adding the `setUpClass` and `tearDownClass`, you will mock asyncio for each test inside you function. The only drawback is that since setUpClass is called once when gerating test, you can't distinguish when and where you mocked function is calle, for example with `self.assertTrue(self.mock_ws.called)`
If you define the same function in `setUp` as defined [here](https://github.com/cnr-ibba/IMAGE-InjectTool/blob/b10ceb20b08eb5009cd29af59265900c3a162508/django-data/image/biosample/tests/common.py#L66), you will setUp the mock object before each method is called, so you can distinguish when you have called your methods.

Reply:
Done